### PR TITLE
DefaultHttp2RemoteFlowController not allocating all available bytes

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
DefaultHttp2RemoteFlowController's allocation algorithm may not allocate all bytes that are available in the connection window. If the 'fair share' based upon weight is not fully used by sibling nodes it was not correctly re-distributed to other sibilings which may be able to utilize part / all of that share.

Modifications:
- Add a unit test which demonstrates the issue.
- Modify the allocation algorithm to ensure all available bytes are allocated.

Result:
Fixes https://github.com/netty/netty/issues/4266

### CODE REVIEW

**Comment by Scottmitch** [comment]
@nmittler - FYI

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java [hunk 1]
```diff
@@ -404,12 +400,22 @@ public boolean visit(Http2Stream child) throws Http2Exception {
...
+                return true;
+            }
+
+            return false;
         }
 
         void feedHungryChildren() throws Http2Exception {
-            if (stillHungry == null) {
-                // There are no hungry children to feed.
+            if (stillHungry == null) { // There are no hungry children to feed.
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: revert this? It was more readable with the carriage return.

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: will do :)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java [hunk 2]
```diff
@@ -404,7 +400,18 @@ public boolean visit(Http2Stream child) throws Http2Exception {
...
             }
 
-            return nextConnectionWindow > 0;
+            if (nextConnectionWindow > 0) {
+                // If this subtree still wants to send then it should be re-considered to take bytes that are unused by
+                // sibling nodes. This is needed because we don't yet know if all the peers will be able to use all of
+                // their "fair share" of the connection window, and if they don't use it then we should divide their
+                // unused shared up for the peers who still want to send.
+                if (state.streamableBytesForTree() > 0) {
+                    stillHungry(child);
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: Don you still need `nextTotalWeight += child.weight`?

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: Oh you updated `stillHungry` to do that ... much better :)

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: I already moved into `stillHungry`. Seemed like it belonged there.

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: > Oh you updated stillHungry to do that ... much better :)

+1. Good suggestion :wink:

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java [hunk 3]
```diff
@@ -756,6 +756,60 @@ public void reprioritizeShouldAdjustOutboundFlow() throws Http2Exception {
     }
 
     /**
+     * Test that the maximum allowed amount the flow controller allows to be sent is always fully allocated if
+     * the streams have at least this much data to send.
```
Comment (nmittler) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java]: attach a link to the bug? Might be helpful in the future to provide context.

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java [hunk 4]
```diff
@@ -756,6 +756,60 @@ public void reprioritizeShouldAdjustOutboundFlow() throws Http2Exception {
...
+     *         0
+     *        /|\
+     *       / | \
+     *      A  B  C
+     *     /
+     *    D
+     * </pre>
+     */
+    @Test
+    public void testAllAvailableChannelBytesAreConsumed() throws Http2Exception {
```
Comment (nmittler) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java]: Maybe something like `unstreamableParentsShouldFeedHungryChildren`?

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java [hunk 5]
```diff
@@ -756,6 +756,60 @@ public void reprioritizeShouldAdjustOutboundFlow() throws Http2Exception {
     }
 
     /**
+     * Test that the maximum allowed amount the flow controller allows to be sent is always fully allocated if
+     * the streams have at least this much data to send.
+     * <pre>
+     *         0
+     *        /|\
+     *       / | \
+     *      A  B  C
```
Comment (nmittler) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java]: Maybe update the diagram to indicate that A, B, and C have zero bytes and D has bytes?

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java (hunk 3 above)
Comment (Scottmitch) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java]: Done

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java [hunk 6]
```diff
@@ -288,7 +288,12 @@ private int maxUsableChannelBytes() {
         return min(connectionState().windowSize(), useableBytes);
     }
 
-    private int writableBytes(int requestedBytes) {
+    /**
+     * Package private for testing purposes only!
+     * @param requestedBytes The desired amount of bytes.
+     * @return The amount of bytes that can be supported by underlying {@link Channel} without queuing "too-much".
+     */
+    final int writableBytes(int requestedBytes) {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: Not sure we need this in the test.  It should be enough to see that D is written to.

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java (hunk 4 above)
Comment (Scottmitch) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java]: sgtm

**Comment by nmittler** [comment]
@Scottmitch great work! Just a few comments.

**Comment by nmittler** [comment]
> +1 for feeding hungry children

@Scottmitch is the Sally Struthers of software :)

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java (hunk 5 above)
Comment (Scottmitch) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java]: How do we indicate this in the diagram? I think the `[B]` notation indicates the connection window is collapsed?

Comment (nmittler) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java]: Start a new notation?  Maybe something like:

```java
            0
       /    |    \
     A(0)  B(0)  C(0)
       |
     D(>0)
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java (hunk 6 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: I added some writabilty change stuff at the last minute that actually wrecked the test. Check updated test.

The key is that we need the streamableBytes to be greater than the amount will will attempt to send. The amount we will be allowed to send is determined by this method.

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java (hunk 5 above)
Comment (Scottmitch) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java]: Sure I can invent something new :)  Wasn't sure if there was an existing notation we were using.

**Comment by nmittler** [comment]
@Scottmitch LGTM ... ship it! :)

**Comment by Scottmitch** [comment]
Cherry-picked 4.1 (93011dd) master (e775eae)

Thanks for review @nmittler! And for being the first person to ever to compare me to Sally Struthers :wink:

**Comment by nmittler** [comment]
@Scottmitch my pleasure! And for only $1/day, you can get me to stop making that comparison ;)

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler lol ... good work :)

**Pull request closed without merging**